### PR TITLE
[#157] Example notebook for NLDI data access

### PR DESCRIPTION
**Pull request by pkdash**

### CODE REVIEW

**Comment by pkdash** [comment]
@thodson-usgs I am waiting for Jeff to do some final edits to the new notebook for nldi data access. After that I it should be ready for review.

**Comment by pkdash** [comment]
@thodson-usgs It is now ready for review. Thanks.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8695bd7a7b335650",
+   "metadata": {},
+   "source": [
+    "### Install the Package\n",
+    "\n",
+    "Use the following code to install the package if it doesn't exist already within your Jupyter Python environment. Note the `nldi` option in the `dataretrieval` package installation. The default `dataretrieval` package installation does not support NLDI data access. You must run the dataretrieval install with the `nldi` option to ensure that you have the necessary capabilities. If you are running this notebook in the CUAHSI JuypyterHub server linked to HydroShare, you will want to run the following pip install command. The base `dataretrieval` package is already installed in the CUAHSI JupyterHub Python environment, but it does not include the NLDI option."
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:24]: I'm not quite sure I follow this. Are you saying the official release of dataretrieval-python does not have the nldi module? I think if you pip install the latest version, the default does come with NLDI. Please let me know if I misunderstand.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+    "\n",
+    "***\n",
+    "\n",
+    "### Examples for the `get_basin()` function:\n",
+    "\n",
+    "The following examples show how to use the `get_basin()` function from the dataretrieval package to retrieve the upstream basin boundary (the contributing watershed for a feature) from the USGS NLDI associated with any feature indexed by NLDI. \n",
+    "\n",
+    "The following arguments are supported:\n",
+    "\n",
+    "* **feature_source** (string): The name of the NLDI feature source.\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:83]: One difficulty I have here is that I don't know all the `feature_source` options. I see in this example `wqp` and the NLDI blog post mentions `nwissite`, but do you know of a place where one can reference all possible inputs?

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+    "* **feature_id** (string): The identifier of the NLDI feature.\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:84]: Similar to previous comment, it's not clear the format of the `feature_id`. I see examples with monitoring locations, like `USGS-#######`, but what about other features?

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "23a84052f0711d2",
+   "metadata": {},
+   "source": [
+    "***\n",
+    "\n",
+    "### Examples for the `get_flowlines()` function:\n",
+    "\n",
+    "The following examples show how to use the `get_flowlines()` function from the dataretrieval package to get flowline data from the USGS NLDI. Flowlines can be traced upsream or downstream from any indexed feature (e.g., a USGS monitoring location), and the result can include mainstem only or tributaries. Flowlines are returned for the specified navigation in WGS84 latitude/longitude coordinats in GeoJSON format.\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb]:
```suggestion
    "The following examples show how to use the `get_flowlines()` function from the dataretrieval package to get flowline data from the USGS NLDI. Flowlines can be traced upstream or downstream from any indexed feature (e.g., a USGS monitoring location), and the result can include mainstem only or tributaries. Flowlines are returned for the specified navigation in WGS84 latitude/longitude coordinats in GeoJSON format.\n",
```

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+    "m"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "42259375160429ab",
+   "metadata": {},
+   "source": [
+    "#### Example 2: Get flowline data using a NHDPlus comid\n",
+    "In some cases, you may wish to get flowline data associated with a NHDPlus COMID rather than flowlines associated with a monitoring station. The following shows how to get flowline data for a COMID as a geopandas dataframe."
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:305]: Can you provide a link to what a COMID is?

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+    "# Display the map\n",
+    "m"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "42259375160429ab",
+   "metadata": {},
+   "source": [
+    "#### Example 2: Get flowline data using a NHDPlus comid\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb]:
```suggestion
    "#### Example 2: Get flowline data using a NHDPlus COMID\n",
```

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+   "source": [
+    "***\n",
+    "\n",
+    "## Usage Examples\n",
+    "\n",
+    "The dataretrieval package provides a number of functions to get data from the USGS NLDI. The following sections provide examples of how each of the available functions can be used\n",
+    "\n",
+    "* `get_basin()`: function used to get the upstream basin boundary for any feature indexed by NLDI.\n",
+    "* `get_flowlines()`: function used to get upstream or downstream flowline data from NLDI.\n",
+    "* `get_features()`: function used to retrieve information about features from any source indexed by NLDI.\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:72]: What constitutes a feature? polygon, point, line?

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 8]
```diff
@@ -0,0 +1,618 @@
...
+    "* **as_json** (boolean): If True, the data will be returned as a python dictionary. If False, the data will be returned as a geopandas dataframe (default is False)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b24a6b1f49ed5f7d",
+   "metadata": {},
+   "source": [
+    "#### Example 1: Get all indexed features along a specified navigation path\n",
+    "Get all of the indexed features from a particular data source using a navigation path that traces upstream along the mainstam (UM) and uses as an origin for the trace a feature from a given feature source (e.g., a monitoring station from the Water Quality Portal). You can get any indexed features from any of the included data sources. Example data sources and the codes you need to retrieve features from those sources include:\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb]:
```suggestion
    "Get all of the indexed features from a particular data source using a navigation path that traces upstream along the mainstem (UM) and uses as an origin for the trace a feature from a given feature source (e.g., a monitoring station from the Water Quality Portal). You can get any indexed features from any of the included data sources. Example data sources and the codes you need to retrieve features from those sources include:\n",
```

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+    "* \"epa_nrsa\" - EPA National Rivers and Streams Assessment\n",
+    "* \"gfv11_pois\" - USGS Geospatial Fabric V1.1 Points of Interest\n",
+    "* \"huc12pp\" - HUC12 Pour Points NHDPlusV2\n",
+    "* \"npdes\" - NPDES Facilities that Discharge to Water\n",
+    "* \"nwisgw\" - NWIS Groundwater Sites\n",
+    "* \"nwissite\" - NWIS Surface Water Sites\n",
+    "* \"WQP\" - Water Quality Portal\n",
+    "* \"comid\" - NHDPlus comid\n",
+    "\n",
+    "There are a few others - for a complete list, consult the getDataSources endpoint of the NLDI API. For this example, we will trace upstream along the mainstem and find any NWIS surface water sites located along the mainstem of the river."
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:379]:
```suggestion
    "There are a few others - for a complete list, consult the getDataSources endpoint of the NLDI API. For this example, we will trace from an NWIS surface water site upstream along the mainstem and find any NWIS surface water sites located along the mainstem of the river."
```

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e6769b9885ef0edb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gdf = nldi.get_features(lat=43.073051, long=-89.401230)\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:495]: What is the buffer distance for a call like this? Can you use navigation mode with a lat/long input?

Comment (ehinman) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:495]: This piece is actually throwing an error for me: ValueError: Error getting features for lat '43.073051' and long '-89.40123'. Error reason: Internal Server Error

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4283a91f6b12446d",
+   "metadata": {},
+   "source": [
+    "***\n",
+    "\n",
+    "### Examples for the `search()` function:\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:506]: I am a little bit confused by the search function.  I don't see how it differs from `get_basin/flowlines/features(as_json=True)`? I could see using `search()` for situations when someone isn't sure what kind of geometry they want, they just want to know the geometries available around a certain point.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb [hunk 12]
```diff
@@ -0,0 +1,618 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b24a6b1f49ed5f7d",
+   "metadata": {},
+   "source": [
+    "#### Example 1: Get all indexed features along a specified navigation path\n",
+    "Get all of the indexed features from a particular data source using a navigation path that traces upstream along the mainstam (UM) and uses as an origin for the trace a feature from a given feature source (e.g., a monitoring station from the Water Quality Portal). You can get any indexed features from any of the included data sources. Example data sources and the codes you need to retrieve features from those sources include:\n",
+    "* \"census2020-nhdpv2\" - 2020 Census Block - NHDPlusV2 Catchment Intersections\n",
```
Comment (ehinman) [inline demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:369]: I think this list is really helpful and should be at the top. Have you considered providing this list as a function like `dataRetrieval::get_nldi_sources()`?

**Review by ehinman — COMMENTED** [review]
This looks pretty good! I definitely learned a lot. As someone who has not used this module before, my questions represent how a beginner might struggle with the nldi functions. I also feel that the nldi module could benefit from more detailed descriptions of what the arguments do/mean. For example, what is `stop_comid` in `get_flowlines()`, and what does `trim_start()` trim? I am also curious if you can use basin polygons as features from which to obtain upstream/downstream basins? If so, that might be a nice example to add.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 5 above)
Comment (thodson-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:305]: @ehinman that's our job to define. I'm also hesitant to use links that we know will break. Perhaps a conceptual description would suffice.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 2 above)
Comment (thodson-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:83]: I think this is something USGS should answer. @dblodgett-usgs might recommend a NLDI API reference.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 1 above)
Comment (thodson-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:24]: NLDI was added after the last official release.

**Comment by thodson-usgs** [comment]
I would just add that @ehinman makes some good points, but some seem out of scope of the PR and will be for us to follow up later.

> I also feel that the nldi module could benefit from more detailed descriptions of what the arguments do/mean. For example, what is `stop_comid` in `get_flowlines()`, and what does `trim_start()` trim?

After reviewing our own swagger doc, I'm also confused. As these functions aren't implemented in the PR, this is a little out of scope and perhaps not something @pkdash can answer.

**Comment by dblodgett-usgs** [comment]
BTW, we are migrating the NLDI to the URL in that link. If the "labs" url is used in this package, can you change "https://labs.waterdata.usgs.gov/api/nldi/" to "https://api.water.usgs.gov/nldi/" throughout?

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 2 above)
Comment (dblodgett-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:83]: A full list is available here: https://api.water.usgs.gov/nldi/linked-data/?f=json 

https://waterdata.usgs.gov/blog/nldi-intro/ has more info

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 3 above)
Comment (dblodgett-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:84]: You have to know the id you are looking for or discover it through navigation requests.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 5 above)
Comment (dblodgett-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:305]: This is the link I usually use. https://www.epa.gov/waterdata/nhdplus-national-hydrography-dataset-plus

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 11 above)
Comment (thodson-usgs) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:506]: I think search mainly allows you to search for multiple feature types in a single query, which can be much more efficient. My recollection was that I wanted to tweak the implementation somewhat, but that's not on @pkdash. He did great for having little-to-no requirements to work from.

File: demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb (hunk 10 above)
Comment (ehinman) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:495]: Disregard bug comment, above.

Comment (pkdash) [reply demos/hydroshare/USGS_dataretrieval_NLDI_Examples.ipynb:495]: @ehinman  I was able to run it without error:
image

**Review by thodson-usgs — APPROVED** [review]
Thanks, for your review @ehinman. Indeed, there are a few loose ends for us to follow up on, but otherwise, this contribution is a clear improvement to our documentation, and I support merging it.

**Pull request merged**